src/tasks/PageIterator.ts

```diff
@@ -0,0 +1,332 @@
+    return this.currentPage?.odataDeltaLink ?? deltaLink;
+  }
+
+  /**
+   * @public
+   * Getter to get the nextLink in the current response
+   * @returns A nextLink which is being used to make requests in future
+   */
+  public getOdataNextLink(): string | undefined {
+    const nextLink = this.currentPage?.["@odata.nextLink"] as string | undefined;
+    return this.currentPage?.odataNextLink ?? nextLink;
+  }
+
+  /**
+   * @public
+   * @async
+   * Iterates over the collection and kicks callback for each item on iteration. Fetches next set of data through nextLink and iterates over again
+   * This happens until the nextLink is drained out or the user responds with a red flag to continue from callback
+   */
+  public async iterate() {
+    const keepIterating = true;
+
+    while (keepIterating) {
+      const advance = this.enumerate();
+      if (!advance) {
+        return;
+      }
+
+      const nextLink = this.getOdataNextLink();
+      if (
+        (nextLink === undefined || nextLink === null || nextLink === "") &&
+        this.cursor >= (this.currentPage?.value.length ?? 0)
+      ) {
+        this.complete = true;
+        this.pagingState = "Complete";
+        return;
+      }
+
+      const nextPage = await this.next();
+      if (!nextPage) {
+        return;
+      }
+      if (!nextPage.odataNextLink && nextPage.odataDeltaLink) {
+        this.pagingState = "Delta";
+      }
+      this.currentPage = nextPage;
+    }
+  }
+
+  /**
+   * @public
+   * Getter to get the state of the iterator
+   */
+  public getPagingState(): PagingState {
+    return this.pagingState;
+  }
+
+  /**
+   * @private
+   * @async
+   * Helper to make a get request to fetch next page with nextLink url and update the page iterator instance with the returned response
+   * @returns A promise that resolves to a response data with next page collection
+   */
+  public async next(): Promise<PageCollection<T> | undefined> {
+    this.pagingState = "InterpageIteration";
+    const requestInformation = new RequestInformation();
+    requestInformation.httpMethod = HttpMethod.GET;
+    requestInformation.urlTemplate = this.getOdataNextLink();
+    if (this.options) {
+      if (this.options.headers) {
+        requestInformation.headers.addAll(this.options.headers);
+      }
+      if (this.options.requestOption) {
+        requestInformation.addRequestOptions(this.options.requestOption);
+      }
+    }
+
+    return await this.requestAdapter.send<PageCollection<T>>(
+      requestInformation,
+      this.parsableFactory,
+      this.errorMappings,
+    );
+  }
+
+  /**
+   * @public
+   * @async
+   * To resume the iteration
+   * Note: This internally calls the iterate method, It's just for more readability.
+   */
+  public async resume() {
+    return this.iterate();
+  }
+
+  /**
+   * @public
+   * To get the completeness status of the iterator
+   * @returns Boolean indicating the completeness
+   */
+  public isComplete(): boolean {
```

Comment:
We can arguably remove this and the boolean property as we are already exposing the paging state.